### PR TITLE
feat(component): new dnd api

### PR DESCRIPTION
**Pull request by EYHN**

### CODE REVIEW

**Comment by EYHN** [comment]
* **#7467** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/7467?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @EYHN and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/frontend/component/src/ui/dnd/draggable.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        if (dragRef.current) {
+          dragRef.current.dataset['capturing'] = 'true';
+        }
+
+        requestAnimationFrame(() => {
+          if (dragRef.current) {
+            delete dragRef.current.dataset['capturing'];
+          }
+        });
+        if (enableCapturingState) {
```
Comment (CatsJuice) [inline packages/frontend/component/src/ui/dnd/draggable.ts]: Isn't this always `true`?

Comment (EYHN) [reply packages/frontend/component/src/ui/dnd/draggable.ts]:
```ts
const { dragRef, capturing } = useDraggable() // enableCapturingState = true
const { dragRef } = useDraggable() // enableCapturingState = false
```

Comment (CatsJuice) [reply packages/frontend/component/src/ui/dnd/draggable.ts]: But `enableCapturingState` is a `MutableRefObject`, shouldn't it be `enableCapturingState.current`?

Comment (EYHN) [reply packages/frontend/component/src/ui/dnd/draggable.ts]: my mistake

File: packages/frontend/component/src/ui/dnd/draggable.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+export const useDraggable = <
+  T extends Record<string, unknown> = Record<string, unknown>,
+>(
+  getOptions: () => DraggableOptions<T> = () => ({}),
+  deps: any[] = []
+) => {
+  const [dragging, setDragging] = useState<boolean>(false);
+
+  // used to capture the dragging state when the drag is started
+  const [capturing, setCapturing] = useState<boolean>(false);
```
Comment (CatsJuice) [inline packages/frontend/component/src/ui/dnd/draggable.ts]: Curious about this state, is it to momentarily apply a special style at the moment drag begins to generate a preview?

Comment (EYHN) [reply packages/frontend/component/src/ui/dnd/draggable.ts]: Yes, when dragging starts, capturing will briefly become true, then become false, and the browser will take a screenshot of the element when capturing = true as a preview of the drag.

File: packages/frontend/component/src/ui/dnd/draggable.ts [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+export type DraggableCustomDragPreviewProps = React.PropsWithChildren<{
+  position?: 'pointer-outside' | 'pointer-center' | 'native';
+}>;
+
+export const useDraggable = <D extends DNDData = DNDData>(
+  getOptions: () => DraggableOptions<D> = () => ({}),
+  deps: any[] = []
+) => {
+  const [dragging, setDragging] = useState<boolean>(false);
+  const [draggingOffset, setDraggingOffset] = useState<{
```
Comment (CatsJuice) [inline packages/frontend/component/src/ui/dnd/draggable.ts]: `draggingPosition` may be more accurate?

File: packages/frontend/component/src/ui/dnd/draggable.ts [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        if (dragRef.current) {
+          dragRef.current.dataset['dragging'] = 'true';
+        }
+      },
+      onDrop: () => {
+        if (enableDragging.current) {
+          setDragging(false);
+        }
+        if (enableDraggingPosition.current) {
+          window.removeEventListener('dragleave', windowEvent.dragleave);
```
Comment (CatsJuice) [inline packages/frontend/component/src/ui/dnd/draggable.ts]: Events are bound to `document.body`

File: packages/frontend/component/src/ui/dnd/draggable.ts [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+          window.removeEventListener('dragover', windowEvent.dragover);
+          setDraggingOffset({
+            offsetX: 0,
+            offsetY: 0,
+            clientX: 0,
```
Comment (CatsJuice) [inline packages/frontend/component/src/ui/dnd/draggable.ts:138]: Initial client position shouldn't be `(0, 0)` ?

Comment (EYHN) [reply packages/frontend/component/src/ui/dnd/draggable.ts:138]: I changed the initial clientXY in onDragStart to real mouse position

**Review by CatsJuice — APPROVED** [review]

**Pull request merged**